framework/sandstone_chrono.cpp

```diff
@@ -11,12 +11,12 @@
 using namespace std;
 using namespace std::chrono;
 
-milliseconds string_to_millisecs(string_view in_string)
+ShortDuration string_to_millisecs(string_view in_string)
 {
     if (in_string.size() == 0)
         return {};
 
-    milliseconds::rep value;
+    ShortDuration::rep value;
     std::from_chars_result r = std::from_chars(in_string.begin(), in_string.end(), value, 10);
     if (r.ec == std::errc{}) {
         string_view suffix = in_string.substr(r.ptr - in_string.data());
```

Comment:
Ehh, yeah I see. We would need `std::chrono`, but with safe conversions between the types.

Reply:
Maybe we should have something like:
```cpp
template <typename Rep1, typename Per1, typename Rep2, typename Per2>
std::expected<std::chrono::duration<Rep1, Per1>, conversion_error_t> 
convert(const std::chrono::duration<Rep2, Per2> &d)
{
    /* Proper conversion based on Per1 / Per2 ratio and width of Rep1 or conversion error */
}
```

Reply:
It can be done, but doing it as efficiently as what we already have is going to require much more code. I've got:
```c++
template <typename To, typename Rep, typename Period> static std::optional<To>
duration_cast_overflow(duration<Rep, Period> from)
{
    using ToRep = typename To::rep;
    using Ratio = std::ratio_divide<Period, typename To::period>;
    static_assert(Ratio::den == 1, "Unsupported conversion, use std::chrono::duration_cast");

    auto value = ToRep(from.count());
    if (value != from.count())
        return std::nullopt;
    if (__builtin_mul_overflow(value, ToRep(Ratio::num), &value))
        return std::nullopt;
    return To(value);
}
...
        std::optional<ShortDuration> result;
        if (suffix == "ms" || suffix.size() == 0)
            result = duration_cast_overflow<ShortDuration>(milliseconds(value));
        else if (suffix == "s")
            result = duration_cast_overflow<ShortDuration>(seconds(value));
        else if (suffix == "m" || suffix == "min")
            result = duration_cast_overflow<ShortDuration>(minutes(value));
        else if (suffix == "h")
            result = duration_cast_overflow<ShortDuration>(hours(value));
```

However, this converts the `int value` to `int64_t` before the call to `duration_cast_overflow`. Fortunately, the compiler is smart enough to realise that the expansion and then contraction are no-ops. The assembly looks like:
```
 146:   0f b6 0e                movzbl (%rsi),%ecx
 149:   80 f9 73                cmp    $0x73,%cl        // 's'
 14c:   74 35                   je     183 <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x183>
 14e:   83 f9 6d                cmp    $0x6d,%ecx       // 'm'
 151:   74 51                   je     1a4 <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x1a4>
 153:   80 f9 68                cmp    $0x68,%cl        // 'h'
 156:   75 1f                   jne    177 <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x177>
 158:   69 c0 80 ee 36 00       imul   $0x36ee80,%eax,%eax
 15e:   0f 80 3a ff ff ff       jo     9e <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x9e>
 164:   5b                      pop    %rbx
 165:   c3                      ret

 183:   69 c0 e8 03 00 00       imul   $0x3e8,%eax,%eax
 189:   0f 80 0f ff ff ff       jo     9e <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x9e>
 18f:   5b                      pop    %rbx
 190:   c3                      ret

 1a4:   69 c0 60 ea 00 00       imul   $0xea60,%eax,%eax
 1aa:   0f 80 ee fe ff ff       jo     9e <string_to_millisecs(std::basic_string_view<char, std::char_traits<char> >)+0x9e>
 1b0:   5b                      pop    %rbx
 1b1:   c3                      ret
```

Want me to replace with this version?

Reply:
Done that.

Reply:
Looks good.

Do you think we could get better code if we had a common, non-templated implementation that takes the `period ratio` (and possibly `Rep`) as runtime arguments and make the template only generate a proper call site for us?
I think this creates a problem with `Rep` ratios not being exactly divisible for integers, therefore require more than one scaling of `Rep`. I feel like there should be a way for using inlining and constant propagation to get the code that we would generate by coding this ourselves and at the same time get the generic source.

And there's still a question if it's even worth the effort.

Reply:
As I understand, the previous version just prepared proper arguments for `imul`, called it and checked for overflow, but now we get many `imul`s.